internal/model/modeldecoder/v2/model.go

```diff
@@ -960,12 +960,54 @@ type log struct {
+	TraceID nullable.String `json:"trace.id" validate:"maxLength=1024"`
+	// TransactionID holds the hex encoded 64 random bits ID of the correlated
+	// transaction.
+	TransactionID nullable.String `json:"transaction.id" validate:"maxLength=1024"`
 	// Message logged as part of the log. In case a parameterized message is
 	// captured, Message should contain the same information, but with any placeholders
 	// being replaced.
 	Message nullable.String `json:"message"`
+	// Level represents the severity of the recorded log.
+	Level nullable.String `json:"log.level" validate:"maxLength=1024"`
+	// Logger represents the name of the used logger instance.
+	Logger nullable.String `json:"log.logger" validate:"maxLength=1024"`
+	// OriginFileName represents the filename containing the sourcecode where the log
+	// originated.
+	OriginFileName nullable.String `json:"log.origin.file.name" validate:"maxLength=1024"`
+	// OriginFileLine represents the line number in the file containing the sourcecode
+	// where the log originated.
+	OriginFileLine nullable.Int `json:"log.origin.file.line"`
+	// OriginFunction represents the function name where the log originated.
+	OriginFunction nullable.String `json:"log.origin.function"`
+	// ErrorType represents the type of the error if the log line represents an error.
+	ErrorType nullable.String `json:"error.type"`
+	// ErrorMessage represents the message contained in the error if the log line
+	// represents an error.
+	ErrorMessage nullable.String `json:"error.message"`
+	// ErrorStacktrace represents the plain text stacktrace of the error the log line
+	// represents.
+	ErrorStacktrace nullable.String `json:"error.stack_trace"`
 	// FAAS holds fields related to Function as a Service events.
 	FAAS faas `json:"faas"`
+	// ServiceName represents name of the service which originated the log line.
+	ServiceName nullable.String `json:"service.name" validate:"maxLength=1024"`
+	// ServiceVersion represents the version of the service which originated the log
+	// line.
+	ServiceVersion nullable.String `json:"service.version" validate:"maxLength=1024"`
+	// ServiceEnvironment represents the environment the service which originated the
+	// log line is running in.
+	ServiceEnvironment nullable.String `json:"service.environment" validate:"maxLength=1024"`
+	// ServiceNodeName represents a unique node name per host for the service which
+	// originated the log line.
+	ServiceNodeName nullable.String `json:"service.node.name" validate:"maxLength=1024"`
+	// ProcessThreadName represents the name of the thread.
+	ProcessThreadName nullable.String `json:"process.thread.name" validate:"maxLength=1024"`
+	// Dataset identifies the source which originated the log line.
+	Dataset nullable.String `json:"dataset" validate:"maxLength=1024"`
+	// Labels are a flat mapping of user-defined key-value pairs.
+	Labels mapstr.M `json:"labels" validate:"inputTypesVals=string;bool;number,maxLengthVals=1024"`
```

Comment:
[**For reviewers**] I have kept the input in flat JSON format, I am not sure if that will work with all ECS logging libraries or if some of the structs need to support both formats (flat and nested)

CC: @felixbarny

Reply:
([Relevant comment](https://github.com/elastic/apm-server/issues/9100#issuecomment-1245207345))

Reply:
Ideally, all properties should be supported both in a flat and a nested notation.

Reply:
@felixbarny Is it correct to assume that a nested notation will have ALL fields as nested. For example:

1. Case 1:
    ```json
    {
        "service": {
            "name": "testsvc",
            "version": "v1.2.0",
            "node.name": "testnode"
        }
    }
    ```
2. Case 2:
    ```json
    {
        "service": {
            "name": "testsvc",
            "version": "v1.2.0",
            "node": {
                "name": "testnode"
            }
        }
    }
    ```
In the above 2 cases, `Case 1` will be incorrect since it has `node.name` in flat notation inside `service` which is in nested notation (the incorrect field will be silently dropped during encoding).

Reply:
Yes, that is correct to assume - a given object path will either be fully nested or fully dotted. However, `log.level` may be dotted while there is also a nested `log` object in the same message. For example:

```json
{
  "log.level": "INFO",
  "log": { "logger": "foo" }
}
```

Reply:
I have updated the code to allow for both nested and flat JSON schema

Reply:
s***********8@gmail.com

https://user-images.githubusercontent.com/115670295/196031034-ea0fbe58-ab63-46bf-bf06-e98a43d7ec76.mp4